man/keyboard.5
.TH KEYBOARD 5 "12 September 1995"
.FB
.SH NAME
keyboard \- \fIFirstBase\fP keyboard map file
.SH SYNOPSIS
.TP
 .firstbase-kbmap
.SH DESCRIPTION
The keyboard map file is used to bind keystroke sequences to
\fIFirstBase\fP actions and intraline editing commands. This map file is used
with
the \fIEditable Input\fP mode within \fIFirstBase\fP.
See \fIsetup\fP(5) and \fIinput\fP(5) on how to enable this input mode.
.PP
When a \fIFirstBase\fP tool is used with \fIEditable Input\fP, the keyboard
map is
loaded from a \fI.firstbase-kbmap\fP file located in either the $HOME
directory,
somewhere along $PATH, or in the \fIFirstBase\fP home directory, in this order.
The first file located is loaded.
.SH KEYBOARD MAP
The \fIFirstBase\fP keyboard map is a list of keyboard keystrokes,
or \fIsequences\fP,
and \fIEditable Input\fP commands, or \fIactions\fP. A \fIsequence\fP is
bound to an \fIaction\fP using the following syntax:
.sp 1
.nf
.nj
.ft CW
      setkey "sequence" action
.ft
.sp 1
.fi
.ju
.PP
The \fBsetkey\fP must be in the first character column \(em these are
the only lines that will be processed. Blanks and tabs are used
to separate the sequence and action parts of each \fIsetkey\fP line.
Comments beyond the action are started using a pound sign (\fB#\fP).
.SH SEQUENCE
Each \fIsequence\fP is a series of characters delimited with
double quotations. These characters (between the quote marks) represent
complete
escape and control keystroke sequences.
.PP
The first character of each sequence must be some kind of control character.
Subsequent characters can be of any type.
Function keys and arrow keys are in this category.
.PP
Control characters are denoted using one of the following:
.sp 1
.nf
.nj
.ft CW
   \\^  .... Control
   \\e  .... Escape
   \\^? .... Del
   \\\\   .... \\
.ft
.sp 1
.fi
.ju
.PP
For example, the sequence \fB\\^f\fP represents <CONTROL>-F, and \fB\\ek\fP
represents <ESCAPE>-k. In these examples, the <CONTROL> character is
depressed while typing the \fBf\fP character, but the <ESCAPE> is a single
keystroke by itself, followed by the single keystroke \fBk\fP.
.PP
A \fIsequence\fP can be as long as needed. Furthermore, multiple sequences
can be bound to the same \fIaction\fP.
.SH ACTION
The \fIaction\fP part of a \fIsetkey\fP line represents a \fIFirstBase\fP
signal or
an intraline editing command. Each action is a single word, or set of
hyphenated words with no blanks. The following is a complete set of all
\fIactions\fP allowed along with a few comments.
.sp 1
.RS 5
.PD 0
.TP 30
beginning-of-line
.TP 30
end-of-line
.TP 30
delete-char-forward
.TP 30
delete-char-backward
.TP 30
delete-to-beginning-of-line
.TP 30
delete-to-end-of-line
.TP 30
move-char-forward
.TP 30
move-char-backward
.TP 30
delete-word-forward
.TP 30
delete-word-backward
.TP 30
move-word-forward
.TP 30
move-word-backward
.TP 30
quit
quit from middle of command sequence
.TP 30
upcase-word
.TP 30
capitalize-word
.TP 30
downcase-word
.TP 30
end
general END signal
.TP 30
help
general HELP signal
.TP 30
abort
general ABORT keystroke
.TP 30
next
next field, database editors
.TP 30
prev
previous field, database editors
.TP 30
redraw
.TP 30
pagedown
database editors - pages, choice pages, etc.
.TP 30
pageup
database editors - pages, choice pages, etc.
.TP 30
bsignal
database editors - backward record
.TP 30
fsignal
database editors - forward record
.TP 30
ssignal
database editors - next matching record
.TP 30
qsignal
database editors - ratio of matching records
.TP 30
writerec
database editors - write record
.TP 30
deleterec
database editors - delete record
.TP 30
system-stop
UNIX stop - .firstbase-init controlled
.TP 30
system-interrupt
UNIX interrupt - sometimes allowed
.PD 0
.RE 5
.sp 1
.SH SAMPLE KEYBOARD MAP
The  following is a sample keyboard map. This map represents the
standard \fIFirstBase\fP keyboard bindings as released with
\fIFirstBase\fP.
.sp 1
.nf
.nj
.ft CW
# intraline editing commands
setkey "\\^a"	beginning-of-line
setkey "\\^e"	end-of-line
setkey "\\^d"	delete-char-forward
setkey "\\^?"	delete-char-backward
setkey "\\^u"	delete-to-beginning-of-line
setkey "\\^k"	delete-to-end-of-line
setkey "\\^f"	move-char-forward
setkey "\\^b"	move-char-backward
setkey "\\ed"	delete-word-forward
setkey "\\e\\^?"	delete-word-backward
setkey "\\^w"	delete-word-backward	# overload function 
setkey "\\ef"	move-word-forward
setkey "\\eb"	move-word-backward
setkey "\\^g"	quit
setkey "\\eu"	upcase-word
setkey "\\ec"	capitalize-word
setkey "\\el"	downcase-word

# standard FirstBase signals
setkey "\\e-"	end
setkey "\\^h"	help
setkey "\\^c"	abort                # general abort keystroke
setkey "\\^n"	next                 # next field, dbedit
setkey "\\^p"	prev                 # prev field, dbedit
setkey "\\^l"	redraw
setkey "\\^v"	pagedown             # dbedit data/choice pages ...
setkey "\\ev"	pageup               # dbedit data/choice pages ...
setkey "\\^x\\^b"	bsignal              # backward record, dbedit
setkey "\\^x\\^f"	fsignal              # forward record, dbedit
setkey "\\^x\\^s"	writerec             # dbedit writerec
setkey "\\^x0"	deleterec            # dbedit delete record

setkey "\\eh"	help

# UNIX system signals
setkey "\\e\\^z"	system-stop          # UNIX stop. setup(5).
setkey "\\e\\^c"	system-interrupt     # UNIX interrupt. sometimes allowed

# Arrow keys - may need tuning to your particular keyboard
setkey "\\e[A"	prev                 # example of an arrow key
setkey "\\e[B"	next                 # example of an arrow key
setkey "\\e[C"	move-char-forward    # example of an arrow key
setkey "\\e[D"	move-char-backward   # example of an arrow key
.ft
.fi
.ju
.sp 1
.SH SPECIAL CASES
A few standard \fIFirstBase\fP keystrokes are overloaded and built-in
to provide default bindings in certain situations. The following will all be
accepted as standard \fIFirstBase\fP signals when no input has been entered
(or altered) and such a signal is allowed.
.sp 1
.PD 0
.TP 10
-
END key. \fIFirstBase\fP standard END keystroke.
.TP 10
^F
fsignal. database editors - forward one record.
.TP 10
^B
bsignal. database editors - backward one record.
.TP 10
^D
dsignal. database editors - autodefault signal.
.TP 10
^Q
qsignal. database editors - report pattern matches.
.TP 10
^S
ssignal. database editors - locate next pattern match.
.TP 10
^E
esignal. dbshell - environment screen command.
.TP 10
^H
help. from first column of an input area.
.PD
.sp 1
.SH SEE ALSO
input(5), setup(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/libinit/keyboard.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: keyboard.c,v 9.1 2001/01/16 02:46:52 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Keyboard_sid[] = "@(#) $Id: keyboard.c,v 9.1 2001/01/16 02:46:52 john Exp $";
#endif

#include <fb.h>
#include <fb_ext.h>
#include <keyboard.h>

/*
 * keyboard - library routines for general mapping of keystrokes to actions.
 *	used by the intraline editing feature of e_fb_input()
 */

typedef struct s_key key;
struct s_key {
   char *k_seq;
   char *k_name;
   int k_len;
   int k_action;
   };

typedef struct s_action action;
struct s_action {
   char *a_name;
   int a_value;
   };

#if FB_PROTOTYPES
static loadkey(char *line);
static action *action_lookup(char *s);
static key *key_signal_lookup(int esig);
static key_string(char *buf, char *seq, int len, int sflag);
#else /* FB_PROTOTYPES */
static void loadkey();
static key_string();
static key *key_signal_lookup();
static action *action_lookup();
#endif /* FB_PROTOTYPES */

extern char *cdb_home;
extern short int cdb_edit_input;

static key keyboard[FB_MAXKEYS];
static int topkey = 0;

char k_quit_char = '\03';

static action action_array[] = {
   "delete-char-backward",	E_DELETE_CHAR_BACKWARD,
   "delete-char-forward",	E_DELETE_CHAR_FORWARD,
   "move-char-backward",	E_MOVE_CHAR_BACKWARD,
   "move-char-forward",		E_MOVE_CHAR_FORWARD,
   "delete-word-backward",	E_DELETE_WORD_BACKWARD,
   "delete-word-forward",	E_DELETE_WORD_FORWARD,
   "move-word-backward",	E_MOVE_WORD_BACKWARD,
   "move-word-forward",		E_MOVE_WORD_FORWARD,
   "beginning-of-line",		E_BEGINNING_OF_LINE,
   "end-of-line",		E_END_OF_LINE,
   "delete-to-beginning-of-line",E_DELETE_TO_BEGINNING_OF_LINE,
   "delete-to-end-of-line",	E_DELETE_TO_END_OF_LINE,
   "capitalize-word",		E_CAPITALIZE_WORD,
   "upcase-word",		E_UPCASE_WORD,
   "downcase-word",		E_DOWNCASE_WORD,
   "dsignal",			E_DSIGNAL,
   "esignal",			E_ESIGNAL,
   "nsignal",			E_ESIGNAL,
   "ysignal",			E_YSIGNAL,
   "psignal",			E_YSIGNAL,
   "redraw",			E_REDRAW,
   "fsignal",			E_FSIGNAL,
   "bsignal",			E_BSIGNAL,
   "ssignal",			E_SSIGNAL,
   "qsignal",			E_QSIGNAL,
   "end",			E_END,
   "abort",			E_ABORT,
   "pageup",			E_PAGEUP,
   "pagedown",			E_PAGEDOWN,
   "next",			E_NEXT,
   "prev",			E_PREV,
   "print",			E_PRINT,
   "default",			E_DEFAULT,
   "clearfield",		E_CLEARFIELD,
   "writerec",			E_WRITEREC,
   "deleterec",			E_DELETEREC,
   "quit",			E_QUIT,
   "help",			E_HELP,
   "system-interrupt",		E_SYS_INTERRUPT,
   "system-stop",		E_SYS_STOP,
   0, 0
   };

/*
 * cdb_init_keyboard - aligns keyboard sequences to actions
 *	opens a HOME .cdb_kbdmap or PATH or CDBHOME ... exactly ONE.
 */

   fb_init_keyboard()
      {
         int fd;
         char buf[FB_MAXLINE], fname[FB_MAXNAME];

         fb_pathname(fname, FB_KBDMAP);
         fd = -1;
         if (fname[0] != NULL)
            fd = open(fname, READ);
         if (fd <= 0){
            strcpy(fname, cdb_home);
            fb_assure_slash(fname);
            strcat(fname, FB_KBDMAP);
            fd = open(fname, READ);
            if (fd <= 0)
               return(FB_ERROR);
            }
         fb_ln_init(fd);
         while (fb_ln_load(buf, FB_MAXLINE) != 0)
            if (strncmp(buf, "setkey", 6) == 0)
               loadkey(buf + 6);
         fb_ln_end();
         return(FB_AOK);
      }

/*
 * loadkey - load up a single <key sequence, action> pair
 */

   static void loadkey(line)
      char *line;

      {
         char nbuf[FB_MAXLINE], *np, *q, *p, kact_buf[FB_MAXNAME];
         int len = 0, i, kact;
         key *k;
         action *a;

         /* skip to first double quote */
         for (p = line; *p && *p != CHAR_QUOTE; p++)
            ;
         if (*p != CHAR_QUOTE)
            return;
         for (p++, np = nbuf; *p; p++, np++){
            if (*p == '\\'){
               /* some meta character */
               p++;
               switch(*p){
                  case 'e':			/* escape */
                  case 'E':
                     *p = '\033';
                     break;
                  case '\\':			/* escaped backslash */
                     *p = '\\';
                     break;
                  case '^':			/* control character */
                     p++;
                     if (islower(*p))
                        *p = toupper(*p);
                     if (*p != '?')
                        *p -= 0100;		/* convert to control */
                     else
                        *p = FB_RUBOUT;
                     break;
                  }
               }
            else if (*p == CHAR_QUOTE)
               break;
            *np = *p;
            len++;
            }
         if (*p != CHAR_QUOTE)
            return;
         strcpy(kact_buf, ++p);
         if ((p = strchr(kact_buf, '#')) != 0)
            *p = NULL;
         fb_rmlead(kact_buf);
         fb_trim(kact_buf);
         i = strlen(kact_buf);
         if (kact_buf[i] == FB_NEWLINE)
            kact_buf[i] = NULL;
         a = action_lookup(kact_buf);
         if (a == NULL)
            return;
         kact = a->a_value;
         k = &keyboard[topkey++];
         k->k_len = len;
         k->k_seq = fb_malloc((unsigned) len);
         k->k_action = kact;
         k->k_name = a->a_name;
         for (np = nbuf, i = 0, q = k->k_seq; i < len; i++)
            *q++ = *np++;
         if (kact == E_QUIT)
            k_quit_char = k->k_seq[0];
      }

/*
 * fb_test_keyboard - return action, 0 if still could be command, -1 if not.
 */

   fb_test_keyboard(s, len)
      char *s;
      int len;

      {
         int i;
         key *k;

         for (i = 0; i < topkey; i++){
            k = &keyboard[i];
            if (k->k_len == len && strncmp(s, k->k_seq, len) == 0)
               return(k->k_action);
            }
         for (i = 0; i < topkey; i++){
            k = &keyboard[i];
            if (k->k_len > len && strncmp(s, k->k_seq, len) == 0)
               return(FB_AOK);
            }
         /* test for internal built ins only if the command is failing */
         if (len == 1){
            switch (*s){
               case FB_DSIGNAL: 	return(E_DSIGNAL);
               case FB_ESIGNAL:
               case FB_ESIGNAL2:
               case FB_ESIGNAL3: 	return(E_ESIGNAL);
               case FB_YSIGNAL:
               case FB_YSIGNAL2: 	return(E_YSIGNAL);
               case FB_KILL:	return(E_ABORT);
               case FB_REDRAW1: 	return(E_REDRAW);
               case FB_FSIGNAL: 	return(E_FSIGNAL);
               case FB_BSIGNAL: 	return(E_BSIGNAL);
               case FB_SSIGNAL: 	return(E_SSIGNAL);
               case FB_QSIGNAL: 	return(E_QSIGNAL);
               case FB_HELPKEY:	return(E_HELP);
               case FB_RUBOUT:	return(E_DELETE_CHAR_BACKWARD);
               }
            }
         return(FB_ERROR);
      }

/*
 * fb_test_help - return 1 if ^H stands for a backward motion, else 0
 */

   fb_test_help()
      {
         char buf[2];
         int f;

         buf[0] = FB_HELPKEY; buf[1] = NULL;
         f = fb_test_keyboard(buf, 1);
         if (f == E_DELETE_CHAR_BACKWARD || f == E_DELETE_WORD_BACKWARD ||
               f == E_MOVE_CHAR_BACKWARD   || f == E_MOVE_WORD_BACKWARD ||
               f == E_HELP)
            return(1);
         return(0);
      }

/*
 * fb_interpret_esig - convert an internal E_SIGNAL into a standard Cdb signal
 */

   fb_interpret_esig(k)
      int k;

      {
         switch(k){
            /* also built in */
            case E_DSIGNAL: 	return(FB_DSIGNAL);
            case E_ESIGNAL: 	return(FB_ESIGNAL);
            case E_YSIGNAL: 	return(FB_YSIGNAL);
            case E_REDRAW: 	return(FB_REDRAW1);
            case E_FSIGNAL: 	return(FB_FSIGNAL);
            case E_BSIGNAL: 	return(FB_BSIGNAL);
            case E_SSIGNAL: 	return(FB_SSIGNAL);
            case E_QSIGNAL: 	return(FB_QSIGNAL);

            /* other signals */
            case E_END:     	return(FB_END);
            case E_ABORT:   	return(FB_ABORT);
            case E_HELP:    	return(FB_QHELP);
            case E_PAGEUP:  	return(FB_PAGEUP);
            case E_PAGEDOWN:	return(FB_PAGEDOWN);
            case E_NEXT:    	return(FB_ESIGNAL);
            case E_PREV:    	return(FB_YSIGNAL);
            case E_PRINT:   	return(FB_PSIGNAL);
            case E_DEFAULT: 	return(FB_DSIGNAL);
            case E_CLEARFIELD:	return(FB_CSIGNAL);
            case E_WRITEREC:	return(FB_WSIGNAL);
            case E_DELETEREC:	return(FB_DELSIGNAL);
            }
         return(0);
      }

/*
 * action_lookup - look up a string given in the action structures.
 */

   static action *action_lookup(s)
      char *s;

      {
         action *a;

         for (a = action_array; a->a_value != 0; a++)
            if (equal(s, a->a_name))
               return(a);
         return(NULL);
      }

/*
 * signal_lookup - look up a signal in the keys, return that *key
 */

   static key *key_signal_lookup(esig)
      int esig;

      {
         int i;
         key *k;

         for (i = 0; i < topkey; i++){
            k = &keyboard[i];
            if (k->k_action == esig)
               return(k);
            }
         return(NULL);
      }

/*
 * fb_key_str - get a printable key string for E_SIGNAL
 */

   fb_key_str(buf, esig)
      char *buf;
      int esig;

      {
         int sflag = 0;
         key *k;

         k = key_signal_lookup(esig);
         if (k != NULL){
            if (k->k_len > 1)
               sflag = 1;
            key_string(buf, k->k_seq, k->k_len, sflag);
            }
      }

/*
 * key_string - convert seq into printable characters for help string
 */

   static key_string(buf, seq, len, sflag)
      char *buf, *seq;
      int len, sflag;

      {
         char c, *p;
         int ctl_flag = 0;

         p = buf;
         *p = NULL;
         for (; len > 0; len--, seq++){
            c = *seq;
            if (c >= 00 && c < 040){
               if (c == 033){
                  strcat(p, "<ESC> ");
                  p += 6;
                  continue;
                  }
               else if (!sflag){
                  strcat(p, "<CTL>-");
                  p += 6;
                  }
               else
                  *p++ = '^';
               c += 0100;
               ctl_flag = 1;
               }
            else if (c == FB_RUBOUT){
               strcat(p, "DEL ");
               p += 4;
               continue;
               }
            *p++ = c;
            if (ctl_flag){
               ctl_flag = 0;
               *p++ = FB_BLANK;
               }
            *p = NULL;
            }
         fb_trim(buf);
      }

/*
 * cdb_seq_convert - convert an old cdb seqence to a new one - for cdbtool.
 */

   fb_seq_convert(buf, seq, len)
      char *buf, *seq;
      int len;

      {
         int esig;
         key *k;

         strncpy(buf, seq, len);
         if (!cdb_edit_input)
            return(len);
         /* first decide what signal we need to gen a seq for */
         if (strncmp(seq, "\016", 1) == 0)		/* ^N */
            esig = E_NEXT;
         else if (strncmp(seq, "\020", 1) == 0)		/* ^P */
            esig = E_PREV;
         else if (strncmp(seq, "\033\020", 2) == 0)	/* ESC ^P */
            esig = E_PAGEUP;
         else if (strncmp(seq, "\033\016", 2) == 0)	/* ESC ^N */
            esig = E_PAGEDOWN;
         else if (strncmp(seq, "\033\027", 2) == 0)	/* ESC ^W */
            esig = E_WRITEREC;
         else if (strncmp(seq, "\033\004", 2) == 0)	/* ESC ^D */
            esig = E_DSIGNAL;
         else if (strncmp(seq, "\006", 1) == 0)		/* ^F */
            esig = E_FSIGNAL;
         else if (strncmp(seq, "\002", 1) == 0)		/* ^B */
            esig = E_BSIGNAL;
         else if (strncmp(seq, "\014", 1) == 0)		/* ^L / ^R */
            esig = E_REDRAW;
         else if (strncmp(seq, "\030", 1) == 0)		/* ^X */
            esig = E_ABORT;
         else
            esig = 0;
         if (esig != 0 && (k = key_signal_lookup(esig)) != NULL){
            strncpy(buf, k->k_seq, k->k_len);
            len = k->k_len;
            }
         return(len);
      }

/*
 * fb_describe_bindings - dump bindings into provided for fs
 */

   fb_describe_bindings(fs)
      FILE *fs;

      {
         int i;
         key *k;
         char rbuf[FB_MAXLINE];

         fprintf(fs, "Key          binding\n");
         fprintf(fs, "---          -------\n\n");
         for (i = 0; i < topkey; i++){
            k = &keyboard[i];
            key_string(rbuf, k->k_seq, k->k_len, 1);
            fprintf(fs, "%-10s   %s\n", rbuf, k->k_name);
            }
      }

/*
 * fb_free_keyboard - free keyboard memory allocations
 */

   fb_free_keyboard()

      {
         int i;
         key *k;

         for (i = 0; i < topkey; i++){
            k = &keyboard[i];
            if (k != NULL && k->k_seq != NULL)
               fb_free(k->k_seq);
            }
      }

#if FB_KEYBOARD_DEBUG
   fb_trace_keyboard()
      {
         int i, j;
         key *k;
         char *q;

         fprintf(stderr, "Begin Trace Keys:\n");
         for (i = 0; i < topkey; i++){
            k = &keyboard[i];
            fprintf(stderr, "Keyboard %d, k_len=%d, k_act=%d: ",
               i, k->k_len, k->k_action);
            for (j = 0, q = k->k_seq; j < k->k_len; j++, q++)
               fprintf(stderr, "\\0%o ", *q);
            fprintf(stderr, "\n");
            }
      }
#endif /* FB_KEYBOARD_DEBUG */
